aten/src/ATen/native/quantized/cpu/ACLUtils.h

```diff
@@ -0,0 +1,257 @@
+#pragma once
+
+#include <ATen/Config.h>
+#if AT_MKLDNN_ACL_ENABLED()
+
+#include <ATen/native/quantized/cpu/OnednnUtils.h>
+#include <arm_compute/core/Error.h>
+#include <arm_compute/core/TensorInfo.h>
+#include <arm_compute/function_info/ActivationLayerInfo.h>
+#include <arm_compute/runtime/Allocator.h>
+#include <arm_compute/runtime/NEON/functions/NEActivationLayer.h>
+#include <arm_compute/runtime/NEON/functions/NEGEMMLowpMatrixMultiplyCore.h>
+#include <arm_compute/runtime/NEON/functions/NEQuantizationLayer.h>
+#include <arm_compute/runtime/Tensor.h>
+#include <array>
+
+namespace at::native::acl_utils {
+
+using ACLDynamicQuantMatmulCacheKey = std::tuple<
+    int64_t, // M
+    bool, // FUSE_RELU
+    int64_t // NUM_THREADS
+    >;
+
+enum class ACLDynamicQuantMatmulCacheKeyIndex {
+  M,
+  FUSE_RELU,
+  NUM_THREADS,
+};
+
+struct ACLDynamicQuantMatmul {
+  arm_compute::Tensor src_s8_tensor;
+  arm_compute::Tensor src_fp32_tensor;
+  arm_compute::Tensor wei_tensor;
+  std::optional<arm_compute::Tensor> bia_tensor;
+  arm_compute::Tensor dst_tensor;
+  arm_compute::NEQuantizationLayer quant;
+  arm_compute::NEGEMMLowpMatrixMultiplyCore gemm;
+  arm_compute::NEActivationLayer acl_relu;
+  // configuration details for the ACL gemm
+  arm_compute::TensorInfo src_s8_tensor_info;
+  arm_compute::TensorInfo src_fp32_tensor_info;
+  arm_compute::TensorInfo wei_tensor_info;
+  std::optional<arm_compute::TensorInfo> bia_tensor_info;
+  arm_compute::TensorInfo dst_tensor_info;
+  arm_compute::GEMMInfo gemm_info;
+  arm_compute::ActivationLayerInfo acl_relu_info{
+      arm_compute::ActivationFunction::RELU};
+
+  // key for use in the cache
+  ACLDynamicQuantMatmulCacheKey key;
+
+  ~ACLDynamicQuantMatmul() {
+    // this will free memory allocated for the quantized src tensor since the
+    // allocation happened through ACL: src_s8_tensor.allocator()->allocate()
+    src_s8_tensor.allocator()->free();
```

Comment:
This is a very unsafe programming model: there are no default constructor, so this structure could be allocated uninitialized, and then freed. I have not tried it, but it's very likely if someone writes something like
```
{
   ACLDynamicQuantMatmul v;
}
```
it will crash in the destructor, as `wei_tensor` has not be allocated, but it's allocator()->free() methods is called unconditionally

Reply:
Actually, sorry for the confusion, my comment above is not right.
`tensor.allocator()->free()` never frees any memory (whether that memory was allocated by ACL or not). It just tells ACL that we're no longer using the pointer - See [here](https://github.com/ARM-software/ComputeLibrary/blob/main/src/runtime/TensorAllocator.cpp#L148) - this can't lead to crashes.

The memory allocated by ACL is freed automatically.

I agree the structure here is not nice.
I added constructors and made sure all memory allocations happen through PyTorch - See [here](https://github.com/pytorch/pytorch/pull/148585/files#diff-db9f299d73216de73f5f8efb40022111c1e26eec780e49dbe70439cb7249013aR59)